### PR TITLE
Refactor impulse response models

### PR DESCRIPTION
**Pull request by maltelueken**
Refactors the impulse response models:

- Moves everything related to impulse response models into a separate module `prfmodel.models.impulse`
- Moves the tests into a separate test module `tests.models.impulse`
- Adds a `ShiftedGammaImpulse` model. This model is simpler than the `TwoGammaImpulse` and has only three parameters (`shift`, `shape`, and `rate`)
- Adds a `ShiftedDerivativeGammaImpulse` model. This model predicts the derivative of the gamma density `f(t)` with respect to `t` which has an overshoot followed by an undershoot. This model is now used as the default in the standard `Gaussian2DPRFModel` because it predicts an undershoot (similar to `TwoGammaImpulse`) but only has three parameters
- Adds an option to all impulse response models (via `BaseImpulse`) to supply a dictionary of default parameters on model object creation. The default parameters will be used during fitting instead of any supplied parameters. In many cases, users do not want to optimize impulse model parameters but just use some sensible default values which is enabled by this option.

### CODE REVIEW

File: src/prfmodel/models/impulse/density.py [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+        The density of the gamma distribution at `value`.
+
+    Notes
+    -----
+    The unnormalized density of the gamma distribution
+    with `shape` :math:`\alpha` and `rate` :math:`\lambda` is given by:
+
+    .. math::
+
+        f(x) = x^{\mathtt{\alpha} - 1} e^{\mathtt{\lambda} x}.
```
Comment (f-hafner) [inline src/prfmodel/models/impulse/density.py]: compared with wikipedia; should this be $x^{\alpha-1} e^{-\lambda x}$?

File: tests/models/impulse/conftest.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+        assert resp.shape == (parameters.shape[0], irf_model.frames.shape[1])
+
+    def test_call_default_parameters(
+        self,
+        irf_model: BaseImpulse,
+        irf_model_default: BaseImpulse,
+        parameters: pd.DataFrame,
+    ):
+        """Test that model response has correct shape."""
```
Comment (f-hafner) [inline tests/models/impulse/conftest.py]:
```suggestion
        """Test that setting default parameters gives a model response with correct shape and works as expected."""
```

File: src/prfmodel/models/impulse/convolve.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    # We pad the pRF response signal on the left side by repeating the first response value
+    # This ensures that, during the convolution, the impulse response starts at every frame of the pRF response
+    # and the output shape is the same as the pRF response shape
+    pad_len = impulse_response.shape[1] - 1
+    response_padding = ops.tile(prf_response[:, :1], (1, pad_len))
+    prf_response_padded = ops.concatenate([response_padding, prf_response], axis=1)
+
+    # Transpose to meet shape requirements of depthwise convolution
+    prf_response_transposed = ops.expand_dims(ops.transpose(prf_response_padded), 0)
+    impulse_response_transposed = ops.expand_dims(ops.transpose(impulse_response), -1)
```
Comment (f-hafner) [inline src/prfmodel/models/impulse/convolve.py]: The current tests for `convolve` only test for shape correctness. I wonder if this pre-processing logic could be packed into its own function, that can then be tested (without testing convolution itself)?  a simple test case would also illustrate better what the pre-processing function does.

File: src/prfmodel/models/gaussian.py [hunk 4]
```diff
@@ -248,9 +248,9 @@ class Gaussian2DPRFModel(SimplePRFModel):
 
     Parameters
     ----------
-    impulse_model : BaseImpulse or type or None, default=TwoGammaImpulse, optional
+    impulse_model : BaseImpulse or type or None, default=ShiftedDerivativeGammaImpulse, optional
         An impulse response model class or instance. Reponse model classes will be instantiated during object
-        initialization. The default creates a `TwoGammaImpulse` instance with default values.
+        initialization. The default creates a `ShiftedDerivativeGammaImpulse` instance with default values.
```
Comment (f-hafner) [inline src/prfmodel/models/gaussian.py]:
```suggestion
        initialization. The default creates a :class:`~prfmodel.models.impulse.ShiftedDerivativeGammaImpulse` instance with default values.
```
Today I learned that sphinx needs this format for the docs to show a clickable link. I think this would fix issue 2 in #39; I'm fine with fixing this systematically across the repo later.

**Review by f-hafner — COMMENTED** [review]
Nice work, well-structured code and tests!

Main suggestions regarding the distribution functions and tests:
- validate the value of the `shift` parameter; test it (similar to the parameters to gamma distribution)
- explore and test edge cases for parameter shapes to shifted (and derivative?) distributions
- test `_shift_density`, for instance with the identity function - to see the padding logic
- consider separating preprocessing logic for convolution and testing that


Also, I think dunder methods are missing, so tagging #42.

File: src/prfmodel/models/impulse/shifted_gamma_deriv.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    >>> impulse_model = ShiftedDerivativeGammaImpulse(
+    >>>     duration=100.0 # 100 seconds
+    >>> )
+    >>> resp = impulse_model(params)
+    >>> print(resp.shape) # (num_rows, duration)
+    (3, 100)
+
+    """
+
+    @property
```
[the remark this answers is not in the document]
Comment (maltelueken) [reply src/prfmodel/models/impulse/shifted_gamma_deriv.py]: Which base class do you mean? Each `BaseImpulse` model can have its own `parameter_names` property, so it cannot be defined there (it only has an abstract property).

File: src/prfmodel/models/impulse/convolve.py (hunk 3 above)
Comment (maltelueken) [reply src/prfmodel/models/impulse/convolve.py]: Yes, I think that is a good idea! I created two private subfunctions with tests that cover the padding and preprocessing respectively. It is still a bit tricky to follow with all the transposing and flipping, but I think it makes it easier to understand what is happening.

File: src/prfmodel/models/gaussian.py (hunk 4 above)
Comment (maltelueken) [reply src/prfmodel/models/gaussian.py]: Thanks, I fixed it here, but, indeed, we need to go through the repo and fix it consistently. Thanks for putting the details into the issue!

File: src/prfmodel/models/impulse/shifted_gamma_deriv.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (f-hafner) [reply src/prfmodel/models/impulse/shifted_gamma_deriv.py]: discussed in person: the option would have been to define a private attribute `_parameter_names` in the subclasses, and define the property in the `BaseImpulse` class, which returns the `_parameter_names`. However, for consistency with other classes, we decided to leave this as it is.
